sagan/train.py

```diff
@@ -0,0 +1,131 @@
+        # Get the interpolated image
+        alpha = tf.random.normal([batch_size, 1, 1, 1], 0.0, 1.0)
+        diff = fake_images - real_images
+        interpolated = real_images + alpha * diff
+
+        with tf.GradientTape() as gp_tape:
+            gp_tape.watch(interpolated)
+            # 1. Get the discriminator output for this interpolated image.
+            pred = self.discriminator(interpolated, training=True)
+
+        # 2. Calculate the gradients w.r.t to this interpolated image.
+        grads = gp_tape.gradient(pred, [interpolated])[0]
+        # 3. Calculate the norm of the gradients.
+        norm = tf.sqrt(tf.reduce_sum(tf.square(grads), axis=[1, 2, 3]))
+        gp = tf.reduce_mean((norm - 1.0) ** 2)
+        return gp
+
+    def train_step(self, real_images):
+        batch_size = tf.shape(real_images)[0]
+
+        for _ in range(2):
+            # Sample random points in the latent space
+            random_latent_vectors = tf.random.normal(
+                shape=(batch_size, self.latent_dim)
+            )
+            generated_images = self.generator(random_latent_vectors)
+            generated_labels = tf.zeros((batch_size, 1))
+
+            # Train the discriminator
+            with tf.GradientTape() as ftape:
+                predictions = self.discriminator(generated_images)
+                d1_loss = self.loss_fn(generated_labels, predictions)
+            grads = ftape.gradient(d1_loss, self.discriminator.trainable_weights)
+            self.d_optimizer.apply_gradients(
+                zip(grads, self.discriminator.trainable_weights)
+            )
+
+            labels = tf.ones((batch_size, 1))
+
+            with tf.GradientTape() as rtape:
+                predictions = self.discriminator(real_images)
+                d_cost = self.loss_fn(labels, predictions)
+
+                # Calculate the gradient penalty
+                gp = self.gradient_penalty(
+                    batch_size, real_images, random_latent_vectors
+                )
+                # Add the gradient penalty to the original discriminator loss
+                d2_loss = d_cost + gp * self.gp_weight
+
+            grads = rtape.gradient(d2_loss, self.discriminator.trainable_weights)
+            self.d_optimizer.apply_gradients(
+                zip(grads, self.discriminator.trainable_weights)
+            )
+
+        ## Train the generator
+        random_latent_vectors = tf.random.normal(shape=(batch_size, self.latent_dim))
+        misleading_labels = tf.ones((batch_size, 1))
+
+        with tf.GradientTape() as gtape:
+            predictions = self.discriminator(self.generator(random_latent_vectors))
+            g_loss = self.loss_fn(misleading_labels, predictions)
+        grads = gtape.gradient(g_loss, self.generator.trainable_weights)
+        self.g_optimizer.apply_gradients(zip(grads, self.generator.trainable_weights))
+
+        return {"d1_loss": d1_loss, "d2_loss": d2_loss, "g_loss": g_loss}
```

Comment:
Return total discriminator loss instead of separtely. I think we should also return the gradient penalty here.